apps/oracles/hyper-borrow-rates/src/providers/euler.rs

```diff
@@ -0,0 +1,160 @@
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+use alloy::primitives::{Address, U256};
+use alloy::providers::ProviderBuilder;
+use alloy::sol_types::SolCall;
+use anyhow::Result;
+use futures::{stream::FuturesUnordered, StreamExt};
+use url::Url;
+
+use crate::domain::{BorrowRateInfo, FeedConfig, RatesPerFeed};
+use crate::providers::onchain::{MyProvider, ETHEREUM_MAINNET_RPC_URL};
+use crate::utils::math::apr_continuous;
+use blocksense_sdk::eth_rpc::eth_call;
+
+pub mod euler {
+    alloy::sol! {
+        #[allow(missing_docs)]
+        #[allow(clippy::too_many_arguments)]
+        #[sol(rpc)]
+        EulerUtilsLens,
+        "src/abi/Euler/utilsLens.json"
+    }
+}
+
+async fn fetch_borrow_rate(
+    provider: &MyProvider,
+    lens_address: Address,
+    vault_address: Address,
+) -> Result<U256> {
+    // let rpc_url = Url::parse(RPC_URL)?;
+    // let provider = ProviderBuilder::new().connect_http(rpc_url.clone());
+
+    let instance = euler::EulerUtilsLens::new(lens_address, provider);
+    let calldata = instance.getAPYs(vault_address).calldata().clone();
+
+    let raw = eth_call(
+        ETHEREUM_MAINNET_RPC_URL,
+        &format!("{:?}", lens_address),
+        &calldata,
+    )
+    .await
+    .map_err(|e| anyhow::Error::msg(format!("{:?}", e)))?;
+
+    let ret = euler::EulerUtilsLens::getAPYsCall::abi_decode_returns(&raw)?;
+
+    Ok(ret.borrowAPY)
+}
+
+pub async fn fetch_borrow_rates_from_euler(
+    feeds_config: Option<&Vec<FeedConfig>>,
+) -> Result<RatesPerFeed> {
+    let mut borrow_rates: RatesPerFeed = RatesPerFeed::new();
+
+    let Some(feeds_config) = feeds_config else {
+        return Ok(borrow_rates);
+    };
+
+    let feeds_by_network = group_feeds_by_network(feeds_config);
+
+    borrow_rates.extend(
+        fetch_borrow_rates_on_ethereum_mainnet(
+            feeds_by_network
+                .get(&SupportedNetworks::EthereumMainnet)
+                .unwrap_or(&Vec::new()),
+        )
+        .await?,
+    );
+    Ok(borrow_rates)
+}
+
+async fn fetch_borrow_rates_on_ethereum_mainnet(feeds: &Vec<FeedConfig>) -> Result<RatesPerFeed> {
```

Comment:
ditto for slices over `Vec` references

```suggestion
async fn fetch_borrow_rates_on_ethereum_mainnet(feeds: &[FeedConfig]) -> Result<RatesPerFeed> {
```